File: src/logging/log.cpp
/*
 * Copyright (C) Canonical, Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <multipass/logging/log.h>

#include <multipass/format.h>

#include <QString>
#include <QtGlobal>

#include <shared_mutex>
#include <stdexcept>

namespace mpl = multipass::logging;

namespace
{
std::shared_timed_mutex mutex;
std::shared_ptr<multipass::logging::Logger> global_logger;

mpl::Level to_level(QtMsgType type)
{
    switch (type)
    {
    case QtDebugMsg:
        return mpl::Level::debug;
    case QtInfoMsg:
        return mpl::Level::info;
    case QtWarningMsg:
        return mpl::Level::warning;
    case QtCriticalMsg:
    case QtFatalMsg:
        return mpl::Level::error;
    }
    throw std::invalid_argument("Unknown Qt log message type");
}

void qt_message_handler(QtMsgType type, const QMessageLogContext&, const QString& message)
{
    auto msg = message.toLocal8Bit();
    mpl::log(to_level(type), "Qt", msg.constData());
}
} // namespace

void mpl::log(Level level, CString category, CString message)
{
    std::shared_lock<decltype(mutex)> lock{mutex};
    if (global_logger)
        global_logger->log(level, category, message);
    else
        fmt::print(stderr, "[{}] [{}] {}\n", as_string(level).c_str(), category.c_str(), message.c_str());
}

mpl::Level mpl::get_logging_level()
{
    if (global_logger)
    {
        return global_logger->get_logging_level();
    }

    return Level::error;
}

void mpl::set_logger(std::shared_ptr<Logger> logger)
{
    std::lock_guard<decltype(mutex)> lock{mutex};
    global_logger = std::move(logger);
    qInstallMessageHandler(qt_message_handler);
}

auto mpl::get_logger() -> Logger* // for tests, don't rely on it lasting
{
    return global_logger.get();
}

File: rxx/rslogger/src/log.rs
/*
 * Copyright (C) Canonical, Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

use crate::ffi;

pub fn log_message(level: ffi::Level, category: String, message: String) {
    match ffi::log_message(level, category, message) {
        Ok(()) => (),
        Err(e) => {
            println!("Log message exception: {e}");
        }
    };
}

pub fn log_message_with_location(
    level: ffi::Level,
    category: String,
    file: &str,
    line: u32,
    module: &str,
    message: String,
) {
    // Equivalent to detail::extract_filename: gets just the file name from the path
    let filename = file
        .split('/')
        .next_back()
        .unwrap_or(file)
        .split('\\')
        .next_back()
        .unwrap_or(file);

    // Format the location info exactly like the C++ counterpart
    let location_message = format!("{}:{}, `{}`: {}", filename, line, module, message);

    log_message(level, category, location_message);
}

#[macro_export]
macro_rules! error {
    ($category:expr, $($arg:tt)+) => {
        $crate::log_message(
            ffi::Level::Error,
            $category.to_string(),
            format!($($arg)+),
        )
    };
}

#[macro_export]
macro_rules! warning {
    ($category:expr, $($arg:tt)+) => {
        $crate::log_message(
            ffi::Level::Warning,
            $category.to_string(),
            format!($($arg)+),
        )
    };
}

#[macro_export]
macro_rules! debug {
    ($category:expr, $($arg:tt)+) => {
        $crate::log_message(
            ffi::Level::Debug,
            $category.to_string(),
            format!($($arg)+),
        )
    };
}

#[macro_export]
macro_rules! trace {
    ($category:expr, $($arg:tt)+) => {
        $crate::log_message(
            ffi::Level::Trace,
            $category.to_string(),
            format!($($arg)+),
        )
    };
}

// --- Location-Aware Macros ---

#[macro_export]
macro_rules! debug_location {
    ($category:expr, $($arg:tt)+) => {
        $crate::log_message_with_location(
            ffi::Level::Debug,
            $category.to_string(),
            file!(),
            line!(),
            module_path!(),
            format!($($arg)+),
        )
    };
}

#[macro_export]
macro_rules! trace_location {
    ($category:expr, $($arg:tt)+) => {
        $crate::log_message_with_location(
            ffi::Level::Trace,
            $category.to_string(),
            file!(),
            line!(),
            module_path!(),
            format!($($arg)+),
        )
    };
}
